examples/linear_deformation.py

```diff
@@ -0,0 +1,1187 @@
+#        return (vectorial_ft_op_inverse(ft_displacement) /
+#                self.par_space[0].cell_volume * 2.0 * np.pi)
+
+
+class L2DataMatchingFunctional(Functional):
+
+    """Basic data matching functional using the L2 norm.
+
+    This functional computes::
+
+        1/2 * ||f - g||_2^2
+
+    for a given element ``g``.
+    """
+
+    def __init__(self, space, data):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        space : `DiscreteLp` with exponent 2.0
+            Space where the data is matched
+        data : `DiscreteLp` element-like
+            Data which is to be matched
+        """
+        if not (isinstance(space, odl.DiscreteLp) and space.exponent == 2.0):
+            raise ValueError('not an L2 space.')
+        super().__init__(space, linear=False)
+        self.data = self.domain.element(data)
+
+    def _call(self, x):
+        """Return ``self(x)``."""
+        return self.domain.dist(x, self.data)
+
+    def gradient(self, x):
+        """Return the gradient in the point ``x``."""
+        return x - self.data
+
+    def derivative(self, x):
+        """Return the derivative in ``x``."""
+        return self.gradient(x).T
+
+
+# Kernel function for any dimensional
+def gauss_kernel(x, sigma):
+    return np.exp(-x ** 2 / (2 * sigma ** 2))
+
+
+# Kernel function
+def kernel(x):
+    scaled = [xi ** 2 / (2 * sigma ** 2) for xi in x]
+    return np.exp(-sum(scaled))
+
+
+def gaussian_kernel_matrix(grid, sigma):
+    """Return the kernel matrix for Gaussian kernel.
+
+    The Gaussian kernel matrix ``K`` in ``n`` dimensions is defined as::
+
+        k_ij = exp(- |x_i - x_j|^2 / (2 * sigma^2))
+
+    where ``x_i, x_j`` runs through all grid points. The final matrix
+    has size ``N x N``, where ``N`` is the total number of grid points.
+
+    Parameters
+    ----------
+    grid : `RegularGrid`
+        Grid where the control points are defined
+    sigma : `float`
+        Width of the Gaussian kernel
+    """
+    point_arrs = grid.points().T
+    matrices = [parr[:, None] - parr[None, :] for parr in point_arrs]
+    for mat in matrices:
+        mat *= mat
+
+    sq_sum = np.sqrt(np.sum(mat for mat in matrices))
+    kernel_matrix = np.exp(-sq_sum / (2 * sigma ** 2))
+    return kernel_matrix
+
+
+def proj_noise(proj_data_shape, mu=0.0, sigma=0.1):
+    """Produce white Gaussian noise for projections of n-D images.
+
+       Produce white Gaussian noise for projections, with the same size
+       as the number of projections.
+
+       Parameters
+       ----------
+       proj_data_shape : shape of the projections
+           Give the size of noise
+       mu : Mean of the norm distribution
+           The defalt is 0.0
+       sigma : Standard deviation of the norm distribution
+           The defalt is 0.1.
+    """
+
+    return np.random.normal(mu, sigma, proj_data_shape)
+
+
+def SNR(signal, noise):
```

Comment:
`def SNR(signal, noise, impl='general'):
    """Compute the signal-to-noise ratio.
    This compute::
        impl='general'
            SNR = s_power / n_power
        impl='dB'
            SNR = 10 * log10 (
                s_power / n_power)
    Parameters
    ----------
    signal : projection
    noise : white noise
    impl : implementation method
    """
    if np.abs(np.asarray(noise)).sum() != 0:
        ave1 = np.sum(signal)/signal.size
        ave2 = np.sum(noise)/noise.size
        s_power = np.sqrt(np.sum((signal - ave1) * (signal - ave1)))
        n_power = np.sqrt(np.sum((noise - ave2) * (noise - ave2)))
        if impl == 'general':
            snr = s_power/n_power
        else:
            snr = 10.0 * np.log10(s_power/n_power)

        return snr

    else:
        return float('inf')`

Two cases are included.